### PR TITLE
Testing: Add type annotations to core/replica_sorter

### PR DESCRIPTION
**Pull request by rdimaio**
Part of #6588

### CODE REVIEW

File: lib/rucio/common/types.py [hunk 1]
```diff
@@ -236,10 +236,10 @@ class TokenValidationDict(TypedDict):
 
 class IPDict(TypedDict):
     ip: str
```
Comment (voetberg) [inline lib/rucio/common/types.py]: Is there a reason for the IP to be required? There's a lot of `ignore`s further down the chain bc of ip

File: lib/rucio/core/replica_sorter.py [hunk 2]
```diff
@@ -228,7 +245,11 @@ def __get_distance_custom(rse: Union[tuple, str], client_location: dict) -> floa
     return cache_val
 
 
-def site_selector(replicas, site, vo):
+def site_selector(
+        replicas: dict[str, Any],
```
Comment (voetberg) [inline lib/rucio/core/replica_sorter.py]: Is there really a dedicated replica dictionary? I think that would make more sense

**Comment by rdimaio** [comment]
Rebased in latest force push

File: lib/rucio/core/replica_sorter.py (hunk 2 above)
Comment (rdimaio) [reply lib/rucio/core/replica_sorter.py]: Addressed in latest revision - created a `ReplicaDict` type

File: lib/rucio/common/types.py (hunk 1 above)
Comment (rdimaio) [reply lib/rucio/common/types.py]: Good point; I've now split this type into two types:
- `IPWithLocationDict` is the one returned by `detect_client_location`, which always returns `ip`, `fqdn` and `site` as `str`
- `IPDict` is used in all other contexts, where any of those values might be `None`

**Comment by rdimaio** [comment]
Latest revision:
- fix merge conflicts

**Review by bari12 — APPROVED** [review]

**Pull request merged**